File: src/main/java/io/github/thebroccolibob/soulorigins/entity/OwnableSkeleton.java
package io.github.thebroccolibob.soulorigins.entity;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.world.World;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

public interface OwnableSkeleton {
    /** In {@link net.minecraft.entity.Entity Entity} */
    World getWorld();
    
    @Nullable UUID soulOrigins$getOwnerUuid();
    void soulOrigins$setOwnerUuid(@Nullable UUID uuid);
    
    default @Nullable PlayerEntity soulOrigins$getOwner() {
        UUID uuid = soulOrigins$getOwnerUuid();
        if (uuid == null) return null;
        return getWorld().getPlayerByUuid(uuid);
    }

    default void soulOrigins$setOwner(PlayerEntity player) {
        soulOrigins$setOwnerUuid(player.getUuid());
    }
    
    default boolean soulOrigins$isTamed() {
        return soulOrigins$getOwnerUuid() != null;
    }
}

File: src/main/kotlin/io/github/thebroccolibob/soulorigins/entity/OwnableSkeleton.kt
package io.github.thebroccolibob.soulorigins.entity

import net.minecraft.entity.mob.AbstractSkeletonEntity
import net.minecraft.entity.player.PlayerEntity

inline var OwnableSkeleton.owner: PlayerEntity?
    get() = `soulOrigins$getOwner`()
    set(player) {
        `soulOrigins$setOwner`(player)
    }

inline val OwnableSkeleton.isTamed
    get() = `soulOrigins$isTamed`()

inline var AbstractSkeletonEntity.owner: PlayerEntity?
    get() = (this as OwnableSkeleton).owner
    set(player) {
        (this as OwnableSkeleton).owner = player
    }

inline val AbstractSkeletonEntity.isTamed
    get() = (this as OwnableSkeleton).isTamed

val a = OwnableSkeleton::owner
